src/box/lua/config/instance_config.lua

```diff
@@ -773,15 +727,152 @@ return schema.new('instance_config', schema.record({
+                        end
+                        assert(uri.password == nil)
+                        if uri.params ~= nil then
+                            w.error('Parameters for iproto.advertise.peer ' ..
+                                    'must be set via ' ..
+                                    'iproto.advertise.peer.params')
+                        end
+                        local ok, err = uri_is_suitable_to_connect(uri)
+                        if not ok then
+                            w.error(err)
+                        end
+                    end,
+                }),
+                params = schema.record({
+                    transport = schema.enum({
+                        'plain',
+                        'ssl',
+                    }),
+                    ssl_ca_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_cert_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_ciphers = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_key_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_password = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_password_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                }),
+            }, {
+                validate = function(data, w)
+                    -- No data in iproto.advertise.peer -- OK.
+                    if next(data) == nil then
+                        return
+                    end
+                    -- If a password is set, a login must also be specified.
+                    if data.password ~= nil and data.login == nil then
+                        w.error('Password cannot be set without setting login')
+                    end
+                    -- If a params is set, an uri must also be specified.
+                    if data.params ~= nil and data.uri == nil then
+                        w.error('Params cannot be set without setting uri')
+                    end
+                end,
             }),
-            sharding = schema.scalar({
-                type = 'string',
-                default = box.NULL,
-                validate = advertise_peer_uri_validate,
+            sharding = schema.record({
+                login = schema.scalar({
+                    type = 'string',
+                }),
+                password = schema.scalar({
+                    type = 'string',
+                }),
+                uri = schema.scalar({
+                    type = 'string',
+                    validate = function(data, w)
```

Comment:
Nit: `iproto.advertise.{peer,sharding}` validation function looks the same. Can we reduce the code duplication?

Reply:
Moved this validation to a function.